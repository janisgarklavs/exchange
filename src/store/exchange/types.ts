export interface Exchange {
  base: string;
  date: string;
  rates: Record<string, number>;
}

export type ApiResponse = Record<string, any>;

export enum ExchangeActionTypes {
  FETCH_REQUEST = "@@exchange/FETCH_REQUEST",
  FETCH_SUCCESS = "@@exchange/FETCH_SUCCESS",
  FETCH_ERROR = "@@exchange/FETCH_ERROR"
}

export interface ExchangeState {
  readonly loading: boolean;
  readonly data?: Exchange;
  readonly errors?: string;
}
