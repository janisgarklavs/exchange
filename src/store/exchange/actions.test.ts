import configureMockStore from "redux-mock-store";
import thunk from "redux-thunk";
import fetchMock from "fetch-mock";

import {
  fetchRequest,
  fetchRequestStart,
  fetchSuccess,
  fetchError,
  EXCHANGE_ENDPOINT
} from "./actions";
import { Exchange, ExchangeActionTypes } from "./types";

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

const store = mockStore();

describe("exchange actions", () => {
  beforeEach(() => {
    store.clearActions();
  });
  describe("fetchRequestStart", () => {
    test("Dispatches the correct action and payload", () => {
      store.dispatch(fetchRequestStart());
      expect(store.getActions()).toMatchSnapshot();
    });
  });
  describe("fetchSuccess", () => {
    test("Dispatches the correct action and payload", () => {
      const data = <Exchange>{};
      store.dispatch(fetchSuccess(data));
      expect(store.getActions()).toMatchSnapshot();
    });
  });
  describe("fetchError", () => {
    test("Dispatches the correct action and payload", () => {
      store.dispatch(fetchError("something failed"));
      expect(store.getActions()).toMatchSnapshot();
    });
  });

  describe("fetchReqeust", () => {
    test("calls request and success actions on succesfull fetch", done => {
      const data = {
        date: "2019-02-18",
        base: "EUR",
        rates: {
          USD: 1.1328
        }
      };

      fetchMock.mock(EXCHANGE_ENDPOINT, {
        status: 200,
        body: data
      });
      store.dispatch(fetchRequest()).then(() => {
        const expectedActions = store.getActions();
        expect(expectedActions.length).toBe(2);
        expect(expectedActions).toContainEqual({
          type: ExchangeActionTypes.FETCH_REQUEST
        });
        expect(expectedActions).toContainEqual({
          type: ExchangeActionTypes.FETCH_SUCCESS,
          payload: data
        });
        done();
      });
    });

    test("calls request and failure actions on not successful fetch", done => {
      const data = "something funky";
      fetchMock.mock(
        EXCHANGE_ENDPOINT,
        {
          status: 400,
          body: { error: data }
        },
        { overwriteRoutes: true }
      );
      store.dispatch(fetchRequest()).then(() => {
        const expectedActions = store.getActions();
        expect(expectedActions.length).toBe(2);
        expect(expectedActions).toContainEqual({
          type: ExchangeActionTypes.FETCH_REQUEST
        });
        expect(expectedActions).toContainEqual({
          type: ExchangeActionTypes.FETCH_ERROR,
          payload: data
        });
        done();
      });
    });
  });
});
