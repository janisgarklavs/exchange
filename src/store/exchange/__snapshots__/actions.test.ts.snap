// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`exchange actions fetchError Dispatches the correct action and payload 1`] = `
Array [
  Object {
    "meta": undefined,
    "payload": "something failed",
    "type": "@@exchange/FETCH_ERROR",
  },
]
`;

exports[`exchange actions fetchRequestStart Dispatches the correct action and payload 1`] = `
Array [
  Object {
    "meta": undefined,
    "payload": undefined,
    "type": "@@exchange/FETCH_REQUEST",
  },
]
`;

exports[`exchange actions fetchSuccess Dispatches the correct action and payload 1`] = `
Array [
  Object {
    "meta": undefined,
    "payload": Object {},
    "type": "@@exchange/FETCH_SUCCESS",
  },
]
`;
